### PR TITLE
Add `extensions` map for downloaded files.

### PR DESCRIPTION
**Pull request by Jimbly**
Allows specifying the downloaded file's extension when the browser saves a file to disk.  Especially useful for dump files.

### CODE REVIEW

**Comment by acrisci** [comment]
I can see this style of configuration getting out of hand quickly. What if we need to add more configuration to custom fields besides this?

As for the minidump, I think it should have a `.dmp` extension by default.

As for further configuration of custom fields and files, how about using an object for configuration?

```yaml
files:
  - name: customfile
    fileName: customfile.txt
```

Then allow for the simpler configuration by string as well.

I'll have to think about the semantics of that, but it seems to be a better option given that more configuration might be in order.

**Comment by Jimbly** [comment]
Sounds like a reasonable idea.

Already I want to add a formatting function to configure how one of my custom fields is displayed (it's a Yaml block of a bunch of key/value pairs), though I'm not sure how that should be done - maybe the config should reference a node module/.js file which is require()'d to do the formatting, or could have some built-in formatters, but extendable is probably better, though too extendable lies the way of madness =).

**Comment by Jimbly** [comment]
Okay, I took a stab at this.  customFields options are now all objects, I added some post-processing to the config to normalize things, and moved the `ip` and `upload_file_minidump` fields into the config object for less custom code in crashreport.coffee's SQL schema setup, and to apply a default downloadAs parameter for the minidump.  If we want to add formatting methods here (maybe a Handlebars snippet would be a reasonable way to do that...) for the `params`, we could do that as well.  I was thinking adding a formatter for IP would be useful (include a geoip link).

Though I've got tons of Node.js experience, this is the first time I've done anything on Coffeescript, so if there's anything that's got a cleaner way to do it in Coffeescript, feel free to let me know =)

File: src/app.coffee [hunk 1]
```diff
@@ -285,11 +285,17 @@ run = ->
...
+      contents = crashreport.get(field)
 
       if not Buffer.isBuffer(contents)
         return res.status(404).send 'Crash report field is not a file'
 
+      # Find appropriate downloadAs file name
+      filename = config.get("customFields:filesById:#{field}:downloadAs") || field
+      filename = filename.replace('#{id}', req.params.id)
+
+      res.setHeader('content-disposition', "attachment; filename=\"#{filename}\"");
```
Comment (acrisci) [inline src/app.coffee]: You don't need a semicolon here. Use `grunt lint` to see linting errors.

File: views/symfile-index.handlebars [hunk 2]
```diff
@@ -26,6 +26,6 @@
 </table>
 {{> pagination}}
 {{else}}
-  <p>No crash reports</p>
+  <p>No symbol files</p>
```
Comment (acrisci) [inline views/symfile-index.handlebars]: Put this in a separate pull request. Use `git add -p` for interactive adds.

File: README.md [hunk 3]
```diff
@@ -105,7 +105,10 @@ database:
   storage: '/home/myuser/.simple-breakpad-server/database.sqlite'
   logging: false
 customFields:
-  files: ['customfile']
+  files:
+    - 'customfile1'
```
Comment (acrisci) [inline README.md]: I think it's kind of confusing to document both ways of adding a custom field in the README. Let's keep one canonical way of adding a custom field and keep the old way for backwards compatibility only.

File: src/config.coffee [hunk 4]
```diff
@@ -36,6 +36,40 @@ nconf.defaults
...
 
+# Post-process custom files and params
+customFields = nconf.get('customFields')
+
+# Ensure array
+customFields.files = customFields.files || []
+# Always add upload_file_minidump file as first file
+customFields.files.splice(0, 0,
+  name: 'upload_file_minidump'
+  downloadAs: 'upload_file_minidump.#{id}.dmp'
```
Comment (acrisci) [inline src/config.coffee]: If we're going to template this, I think this is the wrong way to go about it. What about using handlebars templates like the views?

Comment (Jimbly) [inline src/config.coffee]: Handlebars seems a little overkill, but would work ^_^.  I needed this so that each download gets a unique name (saves an extra click and/or typing when going through dumps and opening them in the debugger - don't get the overwrite prompt each time, and you can't overwrite a dump that's actively being debugged anyway).  I didn't expect any other fields would want this, but thinking more, they might.  I'm considering throwing in a screenshot capture as one of my custom uploaded files, and it would be useful if I could configure that to save the screenshot right next to the dump for future reference, so might be generally useful.  With handlebars, I think we have to be explicit about what values can be used in the templates - it might be useful to have the other attributes (e.g. be able to add IP, or custom fields like an account name, or crash timestamp, through a template), but maybe just "id" for the express purpose of a unique filename is sufficient for now.

Comment (acrisci) [inline src/config.coffee]: You could create a context with all the non-file parameters and template the string with that. I wouldn't expect someone to use anymore of handlebars than variable expansion, but it would let us expand any variable and maybe even supply helpers if the need arises.

If you don't want to do all this now, that's fine. Just change the template syntax to expand `{{id}}` so we have the option of adding that functionality later.

**Comment by Jimbly** [comment]
Okay, I made the trivial change to the templating to at least look like Handlebars for now, shouldn't be too hard to add in something more complex if anyone has a need for it.

I think that's about it for changes to this branch.  If you don't care about the history, I could rebase this down to a single commit before merging.
